File: src/pipeline/latencyMonitor.ts
import type { PipelineMetrics } from '../types/index.js';
import { config } from '../config/index.js';

/**
 * Tracks latency metrics for the pipeline
 */
class LatencyMonitor {
    private metrics: Map<string, PipelineMetrics[]> = new Map();
    private currentRun: Partial<PipelineMetrics> = {};
    private startTime: number = 0;
    private firstByteTime: number = 0;

    /**
     * Start tracking a new pipeline run
     */
    startRun(runId: string): void {
        this.currentRun = {};
        this.startTime = Date.now();
        this.firstByteTime = 0;
    }

    /**
     * Record Time to First Byte
     */
    recordFirstByte(): void {
        if (this.firstByteTime === 0) {
            this.firstByteTime = Date.now();
            this.currentRun.ttfbMs = this.firstByteTime - this.startTime;
        }
    }

    /**
     * Record a component's latency
     */
    recordLatency(
        component: keyof Omit<PipelineMetrics, 'totalLatencyMs' | 'ttfbMs'>,
        latencyMs: number
    ): void {
        this.currentRun[component] = latencyMs;
    }

    /**
     * End the run and store metrics
     */
    endRun(runId: string): PipelineMetrics {
        const totalLatency = Date.now() - this.startTime;

        const finalMetrics: PipelineMetrics = {
            asrLatencyMs: this.currentRun.asrLatencyMs || 0,
            queryRewriteLatencyMs: this.currentRun.queryRewriteLatencyMs || 0,
            searchLatencyMs: this.currentRun.searchLatencyMs || 0,
            rerankLatencyMs: this.currentRun.rerankLatencyMs || 0,
            llmLatencyMs: this.currentRun.llmLatencyMs || 0,
            voiceOptLatencyMs: this.currentRun.voiceOptLatencyMs || 0,
            ttsLatencyMs: this.currentRun.ttsLatencyMs || 0,
            totalLatencyMs: totalLatency,
            ttfbMs: this.currentRun.ttfbMs || totalLatency,
        };

        // Store metrics
        const runMetrics = this.metrics.get(runId) || [];
        runMetrics.push(finalMetrics);
        this.metrics.set(runId, runMetrics);

        // Log metrics
        this.logMetrics(finalMetrics);

        return finalMetrics;
    }

    /**
     * Log metrics to console
     */
    private logMetrics(metrics: PipelineMetrics): void {
        const ttfbStatus = metrics.ttfbMs <= config.latency.targetTTFB ? '✅' : '⚠️';

        console.log('\n📊 Pipeline Metrics:');
        console.log(`   TTFB: ${metrics.ttfbMs}ms ${ttfbStatus} (target: <${config.latency.targetTTFB}ms)`);
        console.log(`   Total: ${metrics.totalLatencyMs}ms`);
        console.log(`   Breakdown:`);
        console.log(`     - ASR: ${metrics.asrLatencyMs}ms`);
        console.log(`     - Query Rewrite: ${metrics.queryRewriteLatencyMs}ms`);
        console.log(`     - Search: ${metrics.searchLatencyMs}ms`);
        console.log(`     - Rerank: ${metrics.rerankLatencyMs}ms`);
        console.log(`     - LLM: ${metrics.llmLatencyMs}ms`);
        console.log(`     - Voice Opt: ${metrics.voiceOptLatencyMs}ms`);
        console.log(`     - TTS: ${metrics.ttsLatencyMs}ms`);
    }

    /**
     * Get average metrics for a run
     */
    getAverageMetrics(runId: string): PipelineMetrics | null {
        const runMetrics = this.metrics.get(runId);
        if (!runMetrics || runMetrics.length === 0) {
            return null;
        }

        const avg = (key: keyof PipelineMetrics) => {
            return Math.round(
                runMetrics.reduce((sum, m) => sum + m[key], 0) / runMetrics.length
            );
        };

        return {
            asrLatencyMs: avg('asrLatencyMs'),
            queryRewriteLatencyMs: avg('queryRewriteLatencyMs'),
            searchLatencyMs: avg('searchLatencyMs'),
            rerankLatencyMs: avg('rerankLatencyMs'),
            llmLatencyMs: avg('llmLatencyMs'),
            voiceOptLatencyMs: avg('voiceOptLatencyMs'),
            ttsLatencyMs: avg('ttsLatencyMs'),
            totalLatencyMs: avg('totalLatencyMs'),
            ttfbMs: avg('ttfbMs'),
        };
    }

    /**
     * Check if TTFB target is met
     */
    isTTFBMet(): boolean {
        return (this.currentRun.ttfbMs || Infinity) <= config.latency.targetTTFB;
    }

    /**
     * Clear metrics
     */
    clear(): void {
        this.metrics.clear();
        this.currentRun = {};
    }
}

// Singleton instance
let latencyMonitor: LatencyMonitor | null = null;

/**
 * Get the latency monitor singleton
 */
export function getLatencyMonitor(): LatencyMonitor {
    if (!latencyMonitor) {
        latencyMonitor = new LatencyMonitor();
    }
    return latencyMonitor;
}

File: src/pipeline/latency_monitor.py
"""
Voice AI Assistant - Latency Monitor
Tracks pipeline latency metrics for performance optimization
"""

import time
from dataclasses import dataclass, field
from typing import Dict, List, Optional

from src.config import config


@dataclass
class PipelineMetrics:
    """Metrics for a single pipeline run"""
    asr_latency_ms: float = 0
    query_rewrite_latency_ms: float = 0
    search_latency_ms: float = 0
    rerank_latency_ms: float = 0
    llm_latency_ms: float = 0
    voice_opt_latency_ms: float = 0
    tts_latency_ms: float = 0
    total_latency_ms: float = 0
    ttfb_ms: float = 0


class LatencyMonitor:
    """Tracks latency metrics for the pipeline"""
    
    def __init__(self):
        self.metrics: Dict[str, List[PipelineMetrics]] = {}
        self.current_run: PipelineMetrics = PipelineMetrics()
        self.start_time: float = 0
        self.first_byte_time: float = 0
    
    def start_run(self, run_id: str) -> None:
        """Start tracking a new pipeline run"""
        self.current_run = PipelineMetrics()
        self.start_time = time.time()
        self.first_byte_time = 0
    
    def record_first_byte(self) -> None:
        """Record Time to First Byte"""
        if self.first_byte_time == 0:
            self.first_byte_time = time.time()
            self.current_run.ttfb_ms = (self.first_byte_time - self.start_time) * 1000
    
    def record_latency(self, component: str, latency_ms: float) -> None:
        """Record a component's latency"""
        if hasattr(self.current_run, f"{component}_latency_ms"):
            setattr(self.current_run, f"{component}_latency_ms", latency_ms)
    
    def end_run(self, run_id: str) -> PipelineMetrics:
        """End the run and store metrics"""
        self.current_run.total_latency_ms = (time.time() - self.start_time) * 1000
        
        if run_id not in self.metrics:
            self.metrics[run_id] = []
        self.metrics[run_id].append(self.current_run)
        
        # Log metrics
        self._log_metrics(self.current_run)
        
        return self.current_run
    
    def _log_metrics(self, metrics: PipelineMetrics) -> None:
        """Log metrics to console"""
        ttfb_status = '✅' if metrics.ttfb_ms <= config.latency.target_ttfb else '⚠️'
        
        print('\n📊 Pipeline Metrics:')
        print(f'   TTFB: {metrics.ttfb_ms:.0f}ms {ttfb_status} (target: <{config.latency.target_ttfb}ms)')
        print(f'   Total: {metrics.total_latency_ms:.0f}ms')
        print('   Breakdown:')
        print(f'     - ASR: {metrics.asr_latency_ms:.0f}ms')
        print(f'     - Query Rewrite: {metrics.query_rewrite_latency_ms:.0f}ms')
        print(f'     - Search: {metrics.search_latency_ms:.0f}ms')
        print(f'     - Rerank: {metrics.rerank_latency_ms:.0f}ms')
        print(f'     - LLM: {metrics.llm_latency_ms:.0f}ms')
        print(f'     - Voice Opt: {metrics.voice_opt_latency_ms:.0f}ms')
        print(f'     - TTS: {metrics.tts_latency_ms:.0f}ms')
    
    def get_average_metrics(self, run_id: str) -> Optional[PipelineMetrics]:
        """Get average metrics for a run"""
        if run_id not in self.metrics or not self.metrics[run_id]:
            return None
        
        run_metrics = self.metrics[run_id]
        n = len(run_metrics)
        
        avg = PipelineMetrics(
            asr_latency_ms=sum(m.asr_latency_ms for m in run_metrics) / n,
            query_rewrite_latency_ms=sum(m.query_rewrite_latency_ms for m in run_metrics) / n,
            search_latency_ms=sum(m.search_latency_ms for m in run_metrics) / n,
            rerank_latency_ms=sum(m.rerank_latency_ms for m in run_metrics) / n,
            llm_latency_ms=sum(m.llm_latency_ms for m in run_metrics) / n,
            voice_opt_latency_ms=sum(m.voice_opt_latency_ms for m in run_metrics) / n,
            tts_latency_ms=sum(m.tts_latency_ms for m in run_metrics) / n,
            total_latency_ms=sum(m.total_latency_ms for m in run_metrics) / n,
            ttfb_ms=sum(m.ttfb_ms for m in run_metrics) / n,
        )
        
        return avg
    
    def is_ttfb_met(self) -> bool:
        """Check if TTFB target is met"""
        return self.current_run.ttfb_ms <= config.latency.target_ttfb
    
    def clear(self) -> None:
        """Clear all metrics"""
        self.metrics.clear()
        self.current_run = PipelineMetrics()


# Global latency monitor instance
_latency_monitor: Optional[LatencyMonitor] = None


def get_latency_monitor() -> LatencyMonitor:
    """Get the global latency monitor instance"""
    global _latency_monitor
    if _latency_monitor is None:
        _latency_monitor = LatencyMonitor()
    return _latency_monitor
